google/cloud/pubsub/integration_tests/message_integration_test.cc

```diff
@@ -0,0 +1,150 @@
+    ~Cleanup() { action(); }
+  };
+  Cleanup cleanup_topic(
+      [topic_admin, &topic]() mutable { topic_admin.DeleteTopic(topic); });
+
+  auto subscription_metadata = subscription_admin.CreateSubscription(
+      CreateSubscriptionBuilder(subscription, topic));
+  ASSERT_STATUS_OK(subscription_metadata);
+
+  auto publisher =
+      pubsub_internal::CreateDefaultPublisherStub(ConnectionOptions{}, 0);
+  auto subscriber =
+      pubsub_internal::CreateDefaultSubscriberStub(ConnectionOptions{}, 0);
+
+  auto publish = [&]() -> StatusOr<std::vector<std::string>> {
+    grpc::ClientContext context;
+    google::pubsub::v1::PublishRequest request;
+    request.set_topic(topic.FullName());
+    *request.add_messages() = [] {
+      google::pubsub::v1::PubsubMessage m;
+      m.set_message_id("message-0");
+      m.set_data("foo");
+      return m;
+    }();
+    *request.add_messages() = [] {
+      google::pubsub::v1::PubsubMessage m;
+      m.set_message_id("message-1");
+      m.set_data("bar");
+      return m;
+    }();
+    *request.add_messages() = [] {
+      google::pubsub::v1::PubsubMessage m;
+      m.set_message_id("message-2");
+      m.set_data("baz");
+      return m;
+    }();
+
+    auto response = publisher->Publish(context, request);
+    if (!response) return std::move(response).status();
+
+    std::vector<std::string> ids;
+    ids.reserve(response->message_ids_size());
+    for (auto& i : *response->mutable_message_ids()) {
+      ids.push_back(std::move(i));
+    }
+    return ids;
+  };
+
+  auto pull =
+      [&]() -> StatusOr<std::vector<google::pubsub::v1::ReceivedMessage>> {
+    grpc::ClientContext context;
+    google::pubsub::v1::PullRequest request;
+    request.set_subscription(subscription.FullName());
+    request.set_max_messages(5);
+    auto response = subscriber->Pull(context, request);
+    if (!response) return std::move(response).status();
+    std::vector<google::pubsub::v1::ReceivedMessage> messages;
+    messages.reserve(response->received_messages_size());
+    for (auto& m : *response->mutable_received_messages()) {
+      messages.push_back(std::move(m));
+    }
+    return messages;
+  };
+
+  auto ack = [&](std::string const& id) -> Status {
+    grpc::ClientContext context;
+    google::pubsub::v1::AcknowledgeRequest request;
+    request.set_subscription(subscription.FullName());
+    request.add_ack_ids(id);
+    return subscriber->Acknowledge(context, request);
+  };
+
+  auto ids = publish();
+  ASSERT_STATUS_OK(ids);
+
+  while (!ids->empty()) {
+    auto messages = pull();
+    ASSERT_STATUS_OK(messages);
+
+    for (auto const& m : *messages) {
+      SCOPED_TRACE("Search for message " + m.DebugString());
+      auto i = std::find_if(
+          ids->begin(), ids->end(),
+          [&m](std::string const& x) { return m.message().message_id() == x; });
+      EXPECT_STATUS_OK(ack(m.ack_id()));
+      EXPECT_NE(i, ids->end());
+      ids->erase(i);
```

Comment:
`erase(end)` is UB, so we should probably guard this (assuming `ASSERT_NE` is overkill).

Reply:
Doh!  assert might be overkill, we probably want to ack all the other messages anyway.